Additions to ocamltest

richardlford opened the pull request:
This is part 2 of a larger PR, #12085, which includes part 1 (compiler and debugger changes), this part, and a third part which is tests of part 1 that make use of (and hence motivate and test) these changes. Please see that PR for example uses of these changes.

1. Add the ability to have scripts or files with ocamltest variables references expanded. Added a new "expand" action to  camltest. It is mostly like the "copy" action, but the source file is read a line at a time and the lines are expanded. It does not support source directories, but the destination can be a directory.

2. Enhanced ocamltest to have a facility for making built-in functions. These are like ocamltest variables but have a function attached to them, and when the variable is expanded, the arguments are first expanded and then the function is called and its result is returned. Currently they only take one parameter, but it should not be too hard to add the ability to have multiple arguments.

3. Added a new "dumpenv_expanded" action which not only shows the value of each variable but also what they expand to. When variables are originally assigned, their RHS is not expanded, but only later when the variable itself is looked up.

4. Two builtin functions were defined:

- bppm_decode does BUILD_PATH_PREFIX_MAP decoding

- bppm_encode does BUILD_PATH_PREFIX_MAP encoding

See https://reproducible-builds.org/specs/build-path-prefix-map/ and https://github.com/richardlford/build-path-prefix-map-spec/pull/1.

5. Prepare to be able to do Dune tests:

5a. Added an action helper for testing if a program is available in PATH.

5b. Use the helper to make an action, has_dune,
that detects whether dune is available.

richardlford committed:
Additions to ocamltest

This is part 2 of a larger PR, #12085, which includes a part 1
(compiler and debugger changes), this part, and a third part which is
tests of part 1 that make use of (and hence motivate and test) these
changes. Please see that PR for example uses of these changes.

1. Add the ability to have scripts or files with ocamltest variables
references expanded. Added a new "expand" action to ocamltest. It is
mostly like the "copy" action, but the source file is read a line at a
time and the lines are expanded. It does not support source
directories, but the destination can be a directory.

2. Enhanced ocamltest to have a facility for making builtin
functions. These are like ocamltest variables but have a function
attached to them, and when the variable is expanded, the arguments are
first expanded and then the function is called and its result is
returned. Currently they only take one parameter, but it should not be
too hard to add the ability to have multiply arguments.

3. Added a new "dumpenv_expanded" action which not
only shows the value of each variable, but also what
they expand to. When variables are originally
assigned, their RHS is not expanded, but only
later when the variable itself is looked up.

4. Two builtin functions were defined:

- bppm_decode does BUILD_PATH_PREFIX_MAP decoding

- bppm_encode does BUILD_PATH_PREFIX_MAP encoding

See https://reproducible-builds.org/specs/build-path-prefix-map/

5. Prepare to be able to do Dune tests:

5a. Added a action helper for testing if a program
is available in PATH.

5b. Use the helper to make an action, has_dune,
that detects whether dune is available.

richardlford commented:
@shindere, could you please review these changes?

shindere commented:
Richard L Ford (2023/03/21 13:01 -0700):
> @shindere, could you please review these changes?

I will end-up doing so but currently I am very, very busy, sorry.

Please, be sure that (1) I'll do it as soon as I can and (2) it's not
necessary to pingme.

richardlford commented:
> Please, be sure that (1) I'll do it as soon as I can and (2) it's not necessary to pingme.

OK. Thanks.

richardlford commented:
Since I first submitted the PR there have been minor corrections. This now is a squashed and rebased version.

richardlford commented:
@shindere, I see that since I created this PR you created a new test syntax and converted the test base to use it. I like the new syntax and think it is easier to understand. I've converted the tests in this PR to the new syntax (using your handy `-translate` option) and rebased. In addition, for your reference, https://github.com/ocaml/ocaml/pull/12126 makes use of some of the new features that are not covered by these tests.
